fix Dispatcher.php for module frontcontroller call in wrong case name

metacreo opened the pull request:
Fix 500 error output when call module controller in leading Capital. real 'controller.php' call 'Controller'

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop / 8.2.x
| Description?      | if you have /modules/mymodule/controllers/front/somecontroller.php but try to call /module/mymodule/Somecontroller (with leading Cap) you got http 500 in front
| Type?             | bug fix 
| Category?         | FO
| BC breaks?        | no 
| Deprecations?     |  no
| How to test?      | Call any frontcontroller for example /module/cronjobs/Callback instead /module/cronjobs/callback
| UI Tests          | ---
| Fixed issue or discussion?     | ---
| Related PRs       | ---
| Sponsor company   | ---

metacreo committed:
fix Dispatcher.php

Fix 500 error output when call module controller in leading Capital.
real 'controller.php' call 'Controller'

ps-jarvis commented:
<!-- PR_TABLE_DESCRIPTION_ERROR -->
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.2.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))
* The `type` should be one of these: `new feature`, `improvement`, `bug fix` or `refacto`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

ps-jarvis commented:
<!-- PR_WELCOME -->
Hello @metacreo!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

kpodemski reviewed (changes requested):
I don’t see a point in changing that. This has been probably done on purpose to unify the URLs, why won’t you use lowercase path instead?

metacreo commented:
this PR eliminates the 500 possibilities on the site and returns compatibility support
in any case 404 is better than 500
to avoid confusion and support compatibility there should be no conditions only lower case.

_"I don’t see a point in changing that. This has been probably done on purpose to unify the URLs, why won’t you use lowercase path instead?"_

imagine the situations:
1 the module is written on a PC where the case does not matter
2 the module is transferred to a unix system
3 in the module the names of the controllers are deliberately changed
and now all this where robots indexed thousands of pages. Robots hammer the site and get 500 instead of 404 or 301

as always, you don't see the point in doing anything good, no big deal... it will get there in time... as with .htaccess in img dir 🙂

metacreo commented:
@kpodemski is it okay for you to be ability to init a 500 error on a server just from the address bar?

kpodemski commented:
@metacreo 

This code has been there for over 12 years, so I think we should be ok ;-)

Also, this change to keep everything in lowercase was implemented to fix another bug from what I see
https://github.com/PrestaShop/PrestaShop/commit/c3bbcda805654ca9f898819aed89d36abaaba6ff

> if you have /modules/mymodule/controllers/front/somecontroller.php but try to call /module/mymodule/Somecontroller (with leading Cap) you got http 500 in front

I've tried to check:
http://psfixer.test/pl/module/blockwishlist/Lists
http://psfixer.test/pl/module/blockwishlist/LiSTS
and also http://psfixer.test/pl/module/ps_emailalerts/ACCOunt
and none of those URLs returned error

kpodemski commented:
@metacreo 

I checked on MacOS ;-) On one of the production stores, I have it points to the error 500, on the other, it doesn't.

In any case, after your change, we will end up having two URLs pointing to the same resource, right?

> Previously, there was no 500 error in version 1.6 for sure

I don't know. It's been a while since I've used this version. I'm commenting on what I see on Git history, and this code hasn't changed in many years.

metacreo commented:
@kpodemski sorry. just checked on 1.6.1.24 (latest 1.6) and yes, 500 error steel on 1.6 too
https://alpturbo.com/en/module/mailalerts/Account
https://alpturbo.com/en/module/mailalerts/account
so ... its core problem
will be good to fix it
```
            case self::FC_MODULE:
                $module_name = Validate::isModuleName(Tools::getValue('module')) ? Tools::getValue('module') : '';
                $module = Module::getInstanceByName($module_name);
                $controller_class = 'PageNotFoundController';
                if (Validate::isLoadedObject($module) && $module->active) {
                    $controllers = Dispatcher::getControllers(_PS_MODULE_DIR_ . "$module_name/controllers/front/");
                    if (isset($controllers[$this->controller])) {
                        include_once _PS_MODULE_DIR_ . "$module_name/controllers/front/{$this->controller}.php";
                        if (file_exists(
                            _PS_OVERRIDE_DIR_ . "modules/$module_name/controllers/front/{$this->controller}.php"
                        )) {
                            include_once _PS_OVERRIDE_DIR_ . "modules/$module_name/controllers/front/{$this->controller}.php";
                            $controller_class = $module_name . $this->controller . 'ModuleFrontControllerOverride';
                        } else {
                            $controller_class = $module_name . $this->controller . 'ModuleFrontController';
                        }
                    }
                }
```
as we see if no controller case must  trigger `$controller_class = 'PageNotFoundController';`
but trigger error because checks in strtolower.
I think this PR is OK for core fix

kpodemski commented:
I'll be sure to get back to you in a couple of days, I have quite a hectic week, but I get your point about the change. Important is not to have 2 URLs pointing to the same resource IMHO, I agree that having internal server error is far from good.

Eolia commented:
> @kpodemski sorry. just checked on 1.6.1.24 (latest 1.6) and yes, 500 error steel on 1.6 too https://alpturbo.com/en/module/mailalerts/Account https://alpturbo.com/en/module/mailalerts/account so ... its core problem will be good to fix it
> 
> ```
>             case self::FC_MODULE:
>                 $module_name = Validate::isModuleName(Tools::getValue('module')) ? Tools::getValue('module') : '';
>                 $module = Module::getInstanceByName($module_name);
>                 $controller_class = 'PageNotFoundController';
>                 if (Validate::isLoadedObject($module) && $module->active) {
>                     $controllers = Dispatcher::getControllers(_PS_MODULE_DIR_ . "$module_name/controllers/front/");
>                     if (isset($controllers[$this->controller])) {
>                         include_once _PS_MODULE_DIR_ . "$module_name/controllers/front/{$this->controller}.php";
>                         if (file_exists(
>                             _PS_OVERRIDE_DIR_ . "modules/$module_name/controllers/front/{$this->controller}.php"
>                         )) {
>                             include_once _PS_OVERRIDE_DIR_ . "modules/$module_name/controllers/front/{$this->controller}.php";
>                             $controller_class = $module_name . $this->controller . 'ModuleFrontControllerOverride';
>                         } else {
>                             $controller_class = $module_name . $this->controller . 'ModuleFrontController';
>                         }
>                     }
>                 }
> ```
> 
> as we see if no controller case must trigger `$controller_class = 'PageNotFoundController';` but trigger error because checks in strtolower. I think this PR is OK for core fix

This doesn't make sense. In getControllersInDirectory() (which populates the "$controllers" array), all controller names were saved in lowercase (same in 1.6).
So it's logic to compare lowercase with lowercase.
![image](https://github.com/user-attachments/assets/7e189f67-e135-4c5e-a8a0-0c21c1ab47dc)

Hlavtox commented:
It should either throw 404 or just work, whatever way is fine IMHO.

metacreo commented:
@Hlavtox actually this PR leads to exactly this

Hlavtox commented:
@metacreo Ok, but make 1000% sure that you don't breal the scenario when you have capital letter in the URL and the file is Controller.php

Eolia commented:
> 1 the module is written on a PC where the case does not matter

The problem is here.
If the developer does not respect the case he will face many problems on Prestashop.
To avoid your 500 error you can add a check of the existence of the file before including it, but nothing more.

`                    if(isset($controllers[strtolower($this->controller)])
                        && Tools::file_exists_cache(
                            _PS_MODULE_DIR_.$module_name.'/controllers/front/'.$this->controller.'.php'
                    )) {
                        include_once(_PS_MODULE_DIR_.$module_name.'/controllers/front/'.$this->controller.'.php');
                        if(Tools::file_exists_cache(
                            _PS_OVERRIDE_DIR_.'modules/'.$module_name.'/controllers/front/'.$this->controller.'.php'
                        )) {
                            include_once(_PS_OVERRIDE_DIR_.'modules/'.$module_name.'/controllers/front/'.$this->controller.'.php');
                            $controller_class = $module_name.$this->controller.'ModuleFrontControllerOverride';
                        }
                        else {
                            $controller_class = $module_name.$this->controller.'ModuleFrontController';
                        }
                    }
`

metacreo commented:
@Eolia so you suggest forcing everyone to use lower case..? but what about the bunch of modules already written with different cases of controller file names.? among them are native modules

Eolia commented:
No @metacreo . If the file name contains capital letters (Account.php) the code will work.
If it is in lowercase, ditto.
But if it is in lowercase and called in uppercase (or vice versa), with the file_exists you will have a 404 (and not a 500).

Links generated by the module must be case-sensitive for file names.

Eolia commented:
> 3 in the module the names of the controllers are deliberately changed

And also:
- If you change the name of a controller, keep the old one to redirect to it or add a .htaccess.
- Always think about backwards compatibility

Robots will be redirected correctly but also all scripts/themes that could use your controller.

metacreo commented:
@Eolia I don't understand your point.
What's wrong with a simple check on dispatcher level? `if (isset($controllers[$this->controller])) {`
this is first check after getting controller. if it not passed we got 404 by default.

metacreo commented:
> > 3 in the module the names of the controllers are deliberately changed
> 
> And also:
> 
>     * If you change the name of a controller, keep the old one to redirect to it or add a .htaccess.
> 
>     * Always think about backwards compatibility
> 
>     * It's not the core code's job to handle arbitrary changes to modules, really not.
> 
> 
> Robots will be redirected correctly but also all scripts/themes that could use your controller.

@Eolia You are telling me things that I have been doing for over 20 years. Explain your point of view specifically.

Eolia commented:
> @Eolia I don't understand your point. What's wrong with a simple check on dispatcher level? `if (isset($controllers[$this->controller])) {` this is first check after getting controller. if it not passed we got 404 by default.

$controllers has been filled with only lowercase names.
So we have to compare like with like.

> You are telling me things that I have been doing for over 20 years. Explain your point of view specifically.
> At this moment we just talk about 500 prevent. 😀

My point of view is that this control should not be touched (possibly adding the one on the existence of the file which will avoid a 500) and that it is not up to the core to compensate for the lack of rigor of those who modify their modules.

metacreo commented:
> $controllers has been filled with only lowercase names.
Why ?

> So we have to compare like with like.
In this case, we are not talking about comparison, but about the presence of the file.

> My point of view is that this control should not be touched (possibly adding the one on the existence of the file which will avoid a 500) and that it is not up to the core to compensate for the lack of rigor of those who modify their modules.
That is, you propose to complicate the procedure of checking the presence of a file. Add extra code and principle. The result will be the same.

Eolia commented:
@metacreo 
- The filling of the $controllers array in lowercase was decided upon when the Dispatcher class was created, it was probably a choice to harmonize the array.
- This table being filled with a scandir() it can only contain existing files.
- It is clear that Prestashop has not provided for the case where a file name would be modified in terms of its case after installation.
- Your proposal avoids the 500 error but I cannot measure the possible side effects.(I'm trying, huh^^)

Eolia commented:
- Your new file: Account.php file
- In array: $controllers['account.php']
Your file name will never be in $controller.
So you won't get a 500 error but your file will never be called. (Unless you also modify getControllersInDirectory())

metacreo commented:
@Eolia I try to repeat :)
first check of file existences running exactly here. and the most correct solution would be to solve this test right here and forever. especially since the absence (not-found) class is already ready here.
the less and simpler the code, the better and faster it works... you understand. and what side effects can there be here. on the contrary, we eliminate the side effect in the form of 500

Of course you need to test and look for a possible problem, that's why there is such a discussion here. Thank you.

Eolia commented:
And what are you doing here (lines 423 / 443)?

                    $controllers = Dispatcher::getControllers(_PS_MODULE_DIR_ . $tab->module . '/controllers/admin/');
                    **if (!isset($controllers[strtolower($this->controller)])) {**
                        $this->controller = $this->controller_not_found;
                        $controller_class = 'AdminNotFoundController';
                    } else {
                        $controller_name = $controllers[strtolower($this->controller)];
                        // Controllers in modules can be named AdminXXX.php or AdminXXXController.php
                        **include_once _PS_MODULE_DIR_ . "{$tab->module}/controllers/admin/$controller_name.php";**
                        if (file_exists(
                            _PS_OVERRIDE_DIR_ . "modules/{$tab->module}/controllers/admin/$controller_name.php"
                        )) {
                            include_once _PS_OVERRIDE_DIR_ . "modules/{$tab->module}/controllers/admin/$controller_name.php";
                            $controller_class = $controller_name . (
                                strpos($controller_name, 'Controller') ? 'Override' : 'ControllerOverride'
                            );
                        } else {
                            $controller_class = $controller_name . (
                                strpos($controller_name, 'Controller') ? '' : 'Controller'
                            );
                        }
                    }

metacreo commented:
> * Your new file: Account.php file
> 
>     * In array: $controllers['account.php']
>       Your file name will never be in $controllers.
>       So you won't get a 500 error but your file will never be called. (Unless you also modify getControllersInDirectory())
> 
> 
> Am I wrong?

No. Called.
![image](https://github.com/user-attachments/assets/bcbe0f3b-c042-494f-bf1e-67f3a2e9932c)

metacreo committed:
Update PR Fix reverse situation

Case 1: Call 'Account' but controller file is 'account'
Case2: Call 'account' but controller file is 'Account'
Now in both cases 404 if no controller file and works if file exists.

metacreo commented:
@Eolia  you right. I found new 'reverse' issue. Call upper but file lower and otherwise call lower but file upper.
so I just changed isset to array search. Now in both cases we got 404 if cases not match and works if match.

Eolia commented:
@metacreo This approach seems much more relevant to me. The case of the key does not matter, what matters is that the existing file is present in the $controllers array.

Eolia commented:
My personal approach:
- Test the existence of the key
- Get the real name of the file (stored in $controllers)
- Include this file

If you call Account.php and $controllers['account'] exists (Account.php or account.php), you will load the right file.

                if(Validate::isLoadedObject($module) && $module->active) {
                    $controllers = Dispatcher::getControllers(_PS_MODULE_DIR_.$module_name.'/controllers/front/');
                    if(isset($controllers[strtolower($this->controller)])) {
                        $controller_name = $controllers[strtolower($this->controller)];
                        include_once(_PS_MODULE_DIR_.$module_name.'/controllers/front/'.$controller_name.'.php');
                        if(Tools::file_exists_cache(
                            _PS_OVERRIDE_DIR_.'modules/'.$module_name.'/controllers/front/'.$controller_name.'.php'
                        )) {
                            include_once(_PS_OVERRIDE_DIR_.'modules/'.$module_name.'/controllers/front/'.$controller_name.'.php');
                            $controller_class = $module_name.$controller_name.'ModuleFrontControllerOverride';
                        }
                        else {
                            $controller_class = $module_name.$controller_name.'ModuleFrontController';
                        }
                    }
                }

metacreo commented:
@Eolia Why are you trying to cache the file name via the Tools class? Have you done any performance testing in this case? or is this protection against ps bugs when requests are duplicated?

It seems to me that the native PHP function will work much faster than the selection from Tools class :)

I didn't intend to rewrite the function, I just needed to quickly fix the possibility of initiating a 500 error from the address bar.
But since we're on the subject, let's dive deeper. Would this suit you?:

```
                if (Validate::isLoadedObject($module) && $module->active) {
                    $controllers = Dispatcher::getControllers(_PS_MODULE_DIR_ . $module_name . '/controllers/front/');
                    if ($controller_name = $controllers[strtolower($this->controller)] ?? null) {
                        $controller_file = $module_name . '/controllers/front/' . $controller_name . '.php';
                        include_once(_PS_MODULE_DIR_ . $controller_file);
                        $controller_class = $module_name . $controller_name . 'ModuleFrontController';
                        if (file_exists_cache(_PS_OVERRIDE_DIR_ . 'modules/' . $controller_file)) {
                            include_once(_PS_OVERRIDE_DIR_ . 'modules/' . $controller_file);
                            $controller_class = $module_name . $controller_name . 'ModuleFrontControllerOverride';
                        }
                    }
                }
```

Eolia commented:
> Why are you trying to cache the file name via the Tools class? Have you done any performance testing in this case? or is this protection against ps bugs when requests are duplicated?

Yes Tools::file_exists_cache() is not particularly useful here as Dispatcher is only called once per page.

For the rest it seems correct to me.

metacreo commented:
@Eolia  Excepts function file_exists_cache :) 
ok, I will test this several hours on production with high load and change PR. Or if you want open new,

metacreo committed:
Update Dispatcher.php

metacreo reviewed (commented):
Tested
